[WIP] Convert Filebeat's traefik.access to ECS.

webmat opened the pull request:
Caveats:

- Started before `source_ecs` was renamed to `source`
- Can't seem to output user agent parsing results to `user_agent`
- ECS fields defs out of date, will need to update them. So far, noticed:
  - Missing `url.original`
  - Missing `http.request.referrer`

TODO:

- [ ] Finish converting relevant fields to ECS in the processing & expected log
- [ ] Add a test for cases where the remote is a hostname
- [ ] Migrate Kibana objects
- [ ] Migrate ML

Ping @ruflin

ruflin commented on filebeat/_meta/fields.common.yml:
```diff
@@ -95,6 +95,11 @@
     - name: source_ecs
       type: group
       fields:
+      - name: domain
```
Looks like the wrong place for it under `source_ecs`?

webmat replied:
Not sure what you mean. We do have `source.domain` defined in ECS.

Until #8983 is merged, this is the place to define the source's domain :-)

ruflin commented on filebeat/module/traefik/access/config/traefik-access.yml:
```diff
@@ -7,9 +7,9 @@ exclude_files: [".gz$"]
-    tokenizer: '%{traefik.access.remote_ip} - %{traefik.access.user_name} [%{traefik.access.time}]
-    "%{traefik.access.method} %{traefik.access.url} HTTP/%{traefik.access.http_version}"
-    %{traefik.access.response_code} %{traefik.access.message}'
+    tokenizer: '%{traefik.access.remote_ip} - %{user.name} [%{traefik.access.time}]
```
Is remote_ip `source.ip`?

ruflin replied:
is access.time the `@timestamp`?

webmat replied:
It eventually becomes the timestamp, once parsed by IN, yes. 

remote_ip is a host. So in Ingest Node I try to parse it as an IP then store in `source_ecs.ip`, if that fails, it falls back to `source.domain`.

ruflin commented on filebeat/module/traefik/access/config/traefik-access.yml:
```diff
@@ -7,9 +7,9 @@ exclude_files: [".gz$"]
-    %{traefik.access.response_code} %{traefik.access.message}'
+    tokenizer: '%{traefik.access.remote_ip} - %{user.name} [%{traefik.access.time}]
+    "%{http.request.method} %{traefik.access.url} HTTP/%{http.version}"
+    %{http.response.status_code} %{traefik.access.message}'
```
message?

ruflin replied:
@ph Every time I look at dissect I'm so happy that I don't have to look at grok patterns ❤️

webmat replied:
Message here is used as "rest", which gets parsed by GROK ;-) in ingest node

ruflin commented on filebeat/module/traefik/access/ingest/pipeline.json:
```diff
@@ -50,13 +50,22 @@
                 "field": "traefik.access.remote_ip",
-                "target_field": "traefik.access.geoip"
+                "patterns": [
+                    "^(%{IP:source_ecs.ip}|%{HOSTNAME:source_ecs.domain})$"
```
`source` is gone in the Filebeat code, we can map directly to `source.domain`

webmat replied:
Again, waiting for #8983 to be merged.

ruflin commented on filebeat/module/traefik/access/ingest/pipeline.json:
```diff
@@ -50,13 +50,22 @@
+        },
+        {
+            "geoip": {
+                "field": "source_ecs.ip",
```
See above

ruflin commented on filebeat/module/traefik/access/test/test.log:
```diff
@@ -1,2 +1,2 @@
 192.168.33.1 - - [02/Oct/2017:20:22:07 +0000] "GET /ui/favicons/favicon-16x16.png HTTP/1.1" 304 0 "http://example.com/login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36" 262 "Host-host-1" "http://172.19.0.3:5601" 2ms
-85.181.35.98 - - [02/Oct/2017:20:22:08 +0000] "GET /ui/favicons/favicon.ico HTTP/1.1" 304 0 "http://example.com/login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36" 271 "Host-host1 "http://172.19.0.3:5601" 3ms
+85.181.35.98 - - [02/Oct/2017:20:22:08 +0000] "GET /ui/favicons/favicon.ico HTTP/1.1" 304 0 "http://example.com/login" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36" 271 "Host-host1" "http://172.19.0.3:5601" 3ms
```
I'm wondering if this is really a type? Seems to be an awkward place for a copy / paste mistake. Did you check the history? Perhaps we should generate a long log in a follow up PR.

webmat replied:
Traefik's documentation says the frontend host name is quoted. Not sure why it would be quoted in some cases and only half quoted in other cases :-)

tsg commented on filebeat/module/traefik/access/test/test.log-expected.json:
```diff
@@ -3,53 +3,55 @@
         "traefik.access.remote_ip": "85.181.35.98", 
-        "traefik.access.response_code": "304", 
-        "traefik.access.url": "/ui/favicons/favicon.ico", 
         "traefik.access.user_agent.device": "Other", 
```
We have to `user_agent` in ECS? Feels like something fairly important?

@ruflin @webmat A more general comment: For the fields that are not in ECS, but are still present in multiple modules (apache, nginx, traefik, iis, etc.), wouldn't it be better to place them under `http.request`/`http.response` rather than technology-specific fields? That would allow them to be used in "ECS dashboards", otherwise it doesn't work.

ruflin replied:
@webmat Let's sync up on the user_agent issue, not sure I understand the problem.

webmat replied:
I don't either, don't worry 😂

ruflin commented on filebeat/module/traefik/_meta/kibana/6/dashboard/Filebeat-traefik-overview.json:
```diff
@@ -2,597 +2,597 @@
         {
             "attributes": {
-                "description": "", 
+                "description": "",
```
Dashboards should not be updated yet. Can you rever this part?

webmat replied:
done

webmat commented:
Indeed, that was my very first PR. I will revert this.

webmat commented:
@ruflin Renames in Kibana objects and ML are gone now. Should be good for a final review.

I also introduced a few unrelated improvements in ecs-migration.yml here: I had forgotten mentions of `module.user_agent.*` to `user_agent.*`, and I added comments between each module's section. This will really help with future rebases LOL

webmat commented:
Only failure is unrelated. Processor test for field rename, AFAICT: https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-darwin/2020/

webmat commented:
jenkins test this

webmat commented:
This was the very first transition PR I wrote, and it had a lot of nasty history. So the latest push is a full rewrite of the PR on the latest master. This warrants a full review again.

This is also the first access log I touch after ECS Beta 2. As such, it goes a bit farther than the previous access log migrations in migrating to ECS.

Finally, there are 3 new caveats. Things that are missing from this PR, but that I think should be done as follow-up PRs, as they are blocked by various dependencies.

webmat commented:
jenkins, test this

webmat commented:
Failure in Jenkins Filebeat build is caused by the `script.max_compilations_rate` error: [link](https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-linux/beat=filebeat,label=ubuntu/2829/testReport/test_modules/Test/test_fileset_file_7_traefik/)

Seems like the attempt to raise the limit (#9613) is not working. The error still shows the 75/5m default...

Issue to fix is #9587.

Full paste of the error, for posterity:

```
Error Message
not error expected but got: {u'http': {u'version': u'1.0', u'request': {u'method': u'GET'}, u'response': {u'status_code': 200}}, u'log': {u'file': {u'path': u'/go/src/github.com/elastic/beats/filebeat/module/traefik/access/test/test.log'}, u'offset': 1581}, u'url': {u'original': u'/apache_pb.gif'}, u'traefik': {u'access': {u'body_sent': {u'bytes': 2326}, u'user_identifier': u'-'}}, u'@timestamp': u'2000-10-10T20:55:36.000Z', u'read_timestamp': u'2018-12-21T01:39:33.544Z', u'agent': {u'hostname': u'5b3966f996f3', u'type': u'filebeat', u'version': u'7.0.0'}, u'source': {u'ip': u'127.0.0.1', u'address': u'127.0.0.1'}, u'host': {u'name': u'5b3966f996f3'}, u'user': {u'name': u'frank'}, u'error': {u'message': u'[script] Too many dynamic script compilations within, max: [75/5m]; please use indexed, or scripts with parameters instead; this limit can be changed by the [script.max_compilations_rate] setting'}, u'input': {u'type': u'log'}, u'event': {u'module': u'traefik', u'dataset': u'access'}}
-------------------- >> begin captured stdout << ---------------------
Using elasticsearch: http://elasticsearch:9200
Testing traefik/access on /go/src/github.com/elastic/beats/filebeat/tests/system/../../module/traefik/access/test/test.log

--------------------- >> end captured stdout << ----------------------
Stacktrace
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/go/src/github.com/elastic/beats/filebeat/build/python-env/local/lib/python2.7/site-packages/parameterized/parameterized.py", line 392, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/go/src/github.com/elastic/beats/filebeat/tests/system/test_modules.py", line 91, in test_fileset_file
    cfgfile=cfgfile)
  File "/go/src/github.com/elastic/beats/filebeat/tests/system/test_modules.py", line 140, in run_on_file
    obj)
not error expected but got: {u'http': {u'version': u'1.0', u'request': {u'method': u'GET'}, u'response': {u'status_code': 200}}, u'log': {u'file': {u'path': u'/go/src/github.com/elastic/beats/filebeat/module/traefik/access/test/test.log'}, u'offset': 1581}, u'url': {u'original': u'/apache_pb.gif'}, u'traefik': {u'access': {u'body_sent': {u'bytes': 2326}, u'user_identifier': u'-'}}, u'@timestamp': u'2000-10-10T20:55:36.000Z', u'read_timestamp': u'2018-12-21T01:39:33.544Z', u'agent': {u'hostname': u'5b3966f996f3', u'type': u'filebeat', u'version': u'7.0.0'}, u'source': {u'ip': u'127.0.0.1', u'address': u'127.0.0.1'}, u'host': {u'name': u'5b3966f996f3'}, u'user': {u'name': u'frank'}, u'error': {u'message': u'[script] Too many dynamic script compilations within, max: [75/5m]; please use indexed, or scripts with parameters instead; this limit can be changed by the [script.max_compilations_rate] setting'}, u'input': {u'type': u'log'}, u'event': {u'module': u'traefik', u'dataset': u'access'}}
-------------------- >> begin captured stdout << ---------------------
Using elasticsearch: http://elasticsearch:9200
Testing traefik/access on /go/src/github.com/elastic/beats/filebeat/tests/system/../../module/traefik/access/test/test.log

--------------------- >> end captured stdout << ----------------------
Standard Output
Using elasticsearch: http://elasticsearch:9200
Testing traefik/access on /go/src/github.com/elastic/beats/filebeat/tests/system/../../module/traefik/access/test/test.log
```

webmat commented:
jenkins, test this

ruflin commented on filebeat/module/traefik/access/test/test.log-expected.json:
```diff
@@ -2,204 +2,218 @@
+        "user_agent.original": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36", 
+        "user_agent.os.full_name": "Linux", 
+        "user_agent.os.name": "Linux", 
+        "user_agent.os.version": "..", 
```
?

webmat replied:
Yes, this is what the second caveat is about. Right now the UA parser gives us major/minor/patch most of the time, and doesn't give us a full version string.

So I'm setting `user_agent.version` and `user_agent.os.version` ([here](https://github.com/elastic/beats/pull/9005/files#diff-5f2f0d705013b636262f7085a1bd369cR112) and [here](https://github.com/elastic/beats/pull/9005/files#diff-5f2f0d705013b636262f7085a1bd369cR101) respectively) with a trivial interpolation. To actually do this right, ideally IN would give this back to us as a full version string (may even be better wrt test versions like betas), second best approach would be to put together a small reuseable pipeline that rebuilds the version string based on these 3-4 fields that may or may not be there, and leverage that everywhere at once.

And the worst but simplest approach was simply to interpolate directly there. Even if I guard it with an if on the presence of `.major`, we'll have broken-looking version numbers whenever another one of the numbers is missing ([like here](https://github.com/elastic/beats/pull/9005/files#diff-932723c2076c2bd3dd6d24df4628a161R99)). So I didn't want to waste a bunch of time on this.

Another alternative for this PR is simply to address this later, and leave major/minor/patch in place for now, like all of the other access log modules.

ruflin replied:
SGTM to delay this.
